### PR TITLE
test-network: show debug logs of networkd

### PR DESCRIPTION
**Pull request by yuwata**

### CODE REVIEW

**Comment by yuwata** [comment]
@mrc0mmand Do you have any idea why no logs are shown?
```
-- Logs begin at Wed 2019-07-03 22:16:20 BST, end at Thu 2019-07-04 04:20:05 BST. --
-- No entries --
```

**Comment by poettering** [comment]
hmm, ci fails due to timeout. can you re-push maybe? maybe that fixes it...

**Comment by evverx** [comment]
> hmm, ci fails due to timeout.

FWIW @ddstreet opened https://github.com/systemd/systemd/issues/12932 yesterday about TEST-34 getting stuck from time to time.

**Comment by mrc0mmand** [comment]
> @mrc0mmand Do you have any idea why no logs are shown?
> 
> ```
> -- Logs begin at Wed 2019-07-03 22:16:20 BST, end at Thu 2019-07-04 04:20:05 BST. --
> -- No entries --
> ```

Ugh, that's weird, I tried it locally several times and it worked, yet it doesn't show anything on both CentOS and Arch in the CI...

**Comment by mrc0mmand** [comment]
@yuwata Ah, now I see it:

```diff
diff --git a/test/test-network/systemd-networkd-tests.py b/test/test-network/systemd-networkd-tests.py
index b727f00..19134ee 100755
--- a/test/test-network/systemd-networkd-tests.py
+++ b/test/test-network/systemd-networkd-tests.py
@@ -290,6 +290,8 @@ def stop_networkd(show_logs=True, remove_state_files=True):
     check_output('systemctl stop systemd-networkd')
     if show_logs:
         invocation_id = check_output('systemctl show systemd-networkd -p InvocationID --value')
+        print("INVOCATION ID: {}".format(invocation_id))
+        print(check_output("systemctl show systemd-networkd -p InvocationID"))
         print(check_output('journalctl _SYSTEMD_INVOCATION_ID=' + invocation_id))
     if remove_state_files:
         remove_networkd_state_files()

```

```
test_delete_links (__main__.NetworkctlTests) ... 
Found link 26
Found link 25
Found link 24
Found link 2
Found link 1
veth-peer: link is being processed by networkd
veth-peer: link is being processed by networkd
veth-peer: link is being processed by networkd
veth-peer: link is being processed by networkd
veth-peer: link is being processed by networkd
veth-peer: link is being processed by networkd
veth-peer: link is being processed by networkd
veth-peer: link is being processed by networkd
veth-peer: link is being processed by networkd
veth-peer: link is being processed by networkd
veth-peer: link is being processed by networkd
veth-peer: link is being processed by networkd
veth-peer: link is being processed by networkd
veth-peer: link is being processed by networkd
veth-peer: link is being processed by networkd
veth-peer: link is being processed by networkd
veth-peer: link is being processed by networkd
INVOCATION ID: 
InvocationID=
-- Logs begin at Thu 2019-07-04 06:49:51 EDT, end at Thu 2019-07-04 07:37:42 EDT. --
-- No entries --
ok
```

I'm not sure why the InvocationID is not set though.

**Comment by evverx** [comment]
Could you try getting `InvocationID` before `systemd-networkd` is stopped?

**Comment by mrc0mmand** [comment]
```diff
diff --git a/test/test-network/systemd-networkd-tests.py b/test/test-network/systemd-networkd-tests.py
index b727f00..cc11a1b 100755
--- a/test/test-network/systemd-networkd-tests.py
+++ b/test/test-network/systemd-networkd-tests.py
@@ -287,9 +287,12 @@ def remove_networkd_state_files():
         os.remove(os.path.join(networkd_runtime_directory, 'state'))
 
 def stop_networkd(show_logs=True, remove_state_files=True):
+    print("\nBEFORE SYSTEMCTL STOP: " + check_output("systemctl show systemd-networkd -p InvocationID"))
     check_output('systemctl stop systemd-networkd')
     if show_logs:
         invocation_id = check_output('systemctl show systemd-networkd -p InvocationID --value')
+        print("INVOCATION ID: {}".format(invocation_id))
+        print("AFTER SYSTEMCTL STOP: " + check_output("systemctl show systemd-networkd -p InvocationID"))
         print(check_output('journalctl _SYSTEMD_INVOCATION_ID=' + invocation_id))
     if remove_state_files:
         remove_networkd_state_files()
```

```
test_delete_links (__main__.NetworkctlTests) ... 
BEFORE SYSTEMCTL STOP: InvocationID=

Found link 59
Found link 58
Found link 57
Found link 2
Found link 1
test1: link is being processed by networkd
test1: link is being processed by networkd
test1: link is being processed by networkd
test1: link is being processed by networkd
test1: link is being processed by networkd
test1: link is being processed by networkd
test1: link is being processed by networkd
test1: link is being processed by networkd
test1: link is being processed by networkd
test1: link is being processed by networkd
test1: link is being processed by networkd
test1: link is being processed by networkd
test1: link is being processed by networkd
veth99: link is being processed by networkd
veth99: link is being processed by networkd
veth99: link is being processed by networkd
veth99: link is being processed by networkd

BEFORE SYSTEMCTL STOP: InvocationID=16e575c518b74d298e17ad8807c28893
INVOCATION ID: 
AFTER SYSTEMCTL STOP: InvocationID=
-- Logs begin at Thu 2019-07-04 06:49:51 EDT, end at Thu 2019-07-04 07:47:54 EDT. --
-- No entries --
ok
```

**Comment by evverx** [comment]
I'm not sure where the first "BEFORE SYSTEMCTL STOP: InvocationID=" came from but the second one was successful as far as I can tell. Looks like it tried to get the logs when `systemd-networkd` was inactive before it was started for the first time. In theory, the test could keep track of invocation ids by getting them after it starts networkd.

**Comment by mrc0mmand** [comment]
The weird thing is it works on the "vanilla" Fedora Rawhide (i.e. without this PR being compiled & installed there):

```
test_delete_links (__main__.NetworkctlTests) ... 
BEFORE SYSTEMCTL STOP: InvocationID=

Found link 11
Found link 10
Found link 9
Found link 7
Found link 6
Found link 5
Found link 2
Found link 1
still waiting for test1:degraded
still waiting for test1:degraded
still waiting for test1:degraded
still waiting for test1:degraded
still waiting for test1:degraded
still waiting for test1:degraded
still waiting for test1:degraded
still waiting for test1:degraded
still waiting for test1:degraded
still waiting for test1:degraded
still waiting for test1:degraded
still waiting for test1:degraded
still waiting for test1:degraded
still waiting for test1:degraded
still waiting for test1:degraded
still waiting for test1:degraded
still waiting for test1:degraded
still waiting for test1:degraded
still waiting for test1:degraded
Event loop failed: Connection timed out
=== trie on-disk ===
tool version:          241
file size:         9492053 bytes
header size             80 bytes
strings            2069269 bytes
nodes              7422704 bytes
● 11: test1
       Link File: /usr/lib/systemd/network/99-default.link
    Network File: /run/systemd/network/11-dummy.network
            Type: ether
           State: degraded (configured)
          Driver: dummy
      HW Address: 32:84:e9:d5:7e:6c
         Address: fe80::3084:e9ff:fed5:7e6c
=== trie on-disk ===
tool version:          241
file size:         9492053 bytes
header size             80 bytes
strings            2069269 bytes
nodes              7422704 bytes
● 10: veth99
       Link File: /usr/lib/systemd/network/99-default.link
    Network File: /run/systemd/network/netdev-link-local-addressing-yes.network
            Type: ether
           State: degraded (configured)
          Driver: veth
      HW Address: 12:34:56:78:9a:bc
         Address: 169.254.215.59
                  fe80::1034:56ff:fe78:9abc
=== trie on-disk ===
tool version:          241
file size:         9492053 bytes
header size             80 bytes
strings            2069269 bytes
nodes              7422704 bytes
● 9: veth-peer
       Link File: /usr/lib/systemd/network/99-default.link
    Network File: /run/systemd/network/netdev-link-local-addressing-yes.network
            Type: ether
           State: degraded (configured)
          Driver: veth
      HW Address: 12:34:56:78:9a:bd
         Address: 169.254.91.110
                  fe80::1034:56ff:fe78:9abd

BEFORE SYSTEMCTL STOP: InvocationID=1110562efe364aa3bfccd2ae01cee032
INVOCATION ID: 1110562efe364aa3bfccd2ae01cee032
AFTER SYSTEMCTL STOP: InvocationID=1110562efe364aa3bfccd2ae01cee032
-- Logs begin at Wed 2019-03-27 06:59:47 EDT, end at Thu 2019-07-04 08:07:49 EDT. --
Jul 04 08:07:27 host-10-0-136-50 systemd-networkd[1121]: Bus bus-api-network: changing state UNSET → OPENING
Jul 04 08:07:27 host-10-0-136-50 systemd-networkd[1121]: Bus bus-api-network: changing state OPENING → AUTHENTICATING
Jul 04 08:07:27 host-10-0-136-50 systemd-networkd[1121]: timestamp of '/etc/systemd/network' changed
Jul 04 08:07:27 host-10-0-136-50 systemd-networkd[1121]: timestamp of '/run/systemd/network' changed
<..snip...>
Jul 04 08:07:47 host-10-0-136-50 systemd-networkd[1121]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/network1/link/_39 interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=33 reply_cookie=0 signature=sa{sv}as error-name=n/a error-message=n/a
Jul 04 08:07:47 host-10-0-136-50 systemd-networkd[1121]: veth-peer: Link removed
Jul 04 08:07:49 host-10-0-136-50 systemd-networkd[1121]: Bus bus-api-network: changing state RUNNING → CLOSED
ERROR
```
```
# systemctl --version
systemd 241 (v241-2.gita09c170.fc31)
+PAM +AUDIT +SELINUX +IMA -APPARMOR +SMACK +SYSVINIT +UTMP +LIBCRYPTSETUP +GCRYPT +GNUTLS +ACL +XZ +LZ4 +SECCOMP +BLKID +ELFUTILS +KMOD +IDN2 -IDN +PCRE2 default-hierarchy=hybrid
```

This is from the Fedora Rawhide machine with this PR for comparison:
```
# systemctl --version
systemd 242 (242-1150-g60ee589)
+PAM +AUDIT +SELINUX +IMA -APPARMOR +SMACK +SYSVINIT +UTMP +LIBCRYPTSETUP +GCRYPT +GNUTLS +ACL +XZ +LZ4 +SECCOMP +BLKID +ELFUTILS +KMOD -IDN2 +IDN -PCRE2 default-hierarchy=unified
```

**Comment by evverx** [comment]
I think on your Fedora machine systemd either collects "garbage" much slower than on CentOS CI or the test is too slow to get its InvocationID before it is garbage-collected (which seems to be the case I think).

File: test/test-network/systemd-networkd-tests.py [hunk 1]
```diff
@@ -282,17 +282,27 @@ def remove_log_file():
...
         os.remove(os.path.join(networkd_runtime_directory, 'state'))
-        check_output('systemctl start systemd-networkd')
-    else:
-        check_output('systemctl restart systemd-networkd')
+
+def stop_networkd(show_logs=True, remove_state_files=True):
+    check_output('systemctl stop systemd-networkd')
+    if show_logs:
+        invocation_id = check_output('systemctl show systemd-networkd -p InvocationID --value')
+        print(check_output('journalctl _SYSTEMD_INVOCATION_ID=' + invocation_id))
```
Comment (keszybz) [inline test/test-network/systemd-networkd-tests.py:294]: InvocationID should certainly be queried *before* the stop request is enqueued.

**Comment by yuwata** [comment]
@mrc0mmand @evverx and @keszybz Thank you for the help. I do not know why the previous version works fine on my laptop. But anyway, I've force-pushed a revised version, in which the invocation ID is obtained _before_ networkd is stopped.

**Comment by evverx** [comment]
@yuwata it's somewhat tangential to this PR but I think it would be useful if the test checked the status of `networkd` after stopping it. It would help to catch assertions crashing networkd at the end, for example, or issues reported by ASan+UBSan (which cause networkd to exit with a non-zero exit code).

**Comment by yuwata** [comment]
@evverx Yeah, sounds nice. But in later PR.

**Comment by evverx** [comment]
@yuwata sure. Now that this PR is almost merged, should https://github.com/systemd/systemd-centos-ci/blob/aaff9c61a70aa6ff0f5426711320c2579ea4b204/vagrant/vagrant-test-sanitizers.sh#L68 be removed from the script running the test on CentOS? As far as I understand, it doesn't have to look for ASan+UBsan reports in the journal any more.

**Comment by yuwata** [comment]
@evverx I am not sure, but this just make the test script shows journal, not checking ASan or UBsan reports...

**Pull request merged**